stumpy/aampdist.py

```diff
@@ -127,8 +126,7 @@ def aampdist(T_A, T_B, m, percentage=0.05, k=None, p=2.0):
 
     See Section III
     """
-    partial_mp_func = functools.partial(aamp, p=p)
-    return core._mpdist(T_A, T_B, m, partial_mp_func, percentage, k)
+    return core._mpdist(T_A, T_B, m, aamp, percentage, k, p=p)
```

Comment:
I wanted to be consistent. In constrast to `stump`, where we switch to `aamp` when `normalize==False` (and it has its own private functions),  we do not do such thing here. I mean...we use `aamp` here but we pass argument "subseq_isconstant" all the way to "_compute_P_ABBA" and then we decide whether we should use it or not depending on the existence of param `normalize` in the `mp_func`. 

Using `functools.partial` gives me this feeling that we are trying to do the same thing but in two different places: when `normalize==False`, we handle `p` (the arg that is only supported when `normalize==False`) using partial, and then when `normalize==True`, we handle "subseq_isconstans" (the arg that is only supported when `normalize==True`) in `_compute_P_ABBA`.  

Therefore, I thought we could just pass all args and then if normalize, we just use the ones that are only supported when normalize==True, and when normalize==False, we use the arguments that are supported only if `normalize==False`. I wanted to handle both at one place. So, later, if we decide to add more and more arguemtns, we can just pass them all the way to `_compute_P_ABBA` and then separate the normalized / non-normalized paths there at the end.

Reply:
Hmm, I've considered and understand your response but I don't think it is wise to open up `core._mpdist` or `core._compute_P_ABBA` and force those functions to require `normalize`, `p`. At the end of the day, neither of those functions are actually using those parameters themselves (i.e., it's the underlying `mp_func` that would use those parameters) and I think it's better to abstract that away so that those functions are very general. One thing to keep in mind is that `_mpdist` and `_compute_P_ABBA` are both private functions. The more we add parameters like `normalize`, `p`, and other parameters to these functions then we are forced to remember to add them all the way down the chain of function calls. By wrapping them up in `functools.partial` at the highest layer, we simply package it up once and avoid a bunch of `if/else` statements. Now that I think about it more, we could still wrap everything using `functools.partial` and check for the existence of `T_A_subseq_isconstant`/`T_B_subseq_isconstant` inside of `core._compute_P_ABBA` by checking the `keywords` that are wrapped by `functools.partial`:

```
from functools import partial
import inspect
from stumpy import stump

partial_mp_func = partial(stump, T_A_subseq_isconstant=T_A_subseq_isconstant, T_B_subseq_isconstant=T_B_subseq_isconstant)    
```

Then, inside of `core._compute_P_ABBA` (note that we'd remove/omit `T_A_subseq_isconstant`/`T_B_subseq_isconstant` parameters from BOTH `core._mpdist` and `core._compute_P_ABBA`), you'd do:

```
if inspect.signature(mp_func).parameters.get("normalize") is not None:
    # Non-normalized (stump-like)
    # Only non-normalized mp funcs can have a "normalize" parameter in its function signature
    params = partial_mp_func.keywords
    T_A_subseq_isconstant = params.get("T_A_subseq_isconstant")
    T_B_subseq_isconstant = params.get("T_B_subseq_isconstant")
    P_ABBA[: n_A - m + 1] = partial_mp_func(T_A, m, T_B, ignore_trivial=False, T_A_subseq_isconstant=T_A_isconstant, T_B_subseq_isconstant=T_B_subseq_isconstant)[:, 0]
    P_ABBA[n_A - m + 1 :] = partial_mp_func(T_B, m, T_A, ignore_trivial=False, T_B_subseq_isconstant=T_A_isconstant, T_B_subseq_isconstant=T_A_subseq_isconstant)[:, 0]
else:
    # Normalized (aamp-like)
    # Ignore/omit `T_A_subseq_isconstant` and `T_B_subseq_isconstant` parameters for all non-normalized mp funcs 
    P_ABBA[: n_A - m + 1] = partial_mp_func(T_A, m, T_B, ignore_trivial=False)[:, 0]
    P_ABBA[n_A - m + 1 :] = partial_mp_func(T_B, m, T_A, ignore_trivial=False)[:, 0]
``` 

Here we auto-discover the necessary `T_A_subseq_isconstant`/`T_B_subseq_isconstant` if they exist. If not, the default Dict `get` would/should return `None`. And that should be it.

Reply:
Thank you for the thorough explanation. 

> One thing to keep in mind is that _mpdist and _compute_P_ABBA are both private functions. The more we add parameters like normalize, p, and other parameters to these functions then we are forced to remember to add them all the way down the chain of function calls. By wrapping them up in functools.partial at the highest layer, we simply package it up once and avoid a bunch of if/else statements.

Great point! I should keep that in mind! 

> by checking the keywords that are wrapped by functools.partial

Cool! I will revise the code.